systemd/systemd_test.go

```diff
@@ -507,29 +508,113 @@ func (s *SystemdTestSuite) TestLogs(c *C) {
 	c.Check(s.j, Equals, 1)
 }
 
+// mustJSONMarshal panic's if the value cannot be marshaled
+func mustJSONMarshal(v interface{}) *json.RawMessage {
+	b, err := json.Marshal(v)
+	if err != nil {
+		panic(fmt.Sprintf("couldn't marshal json in test fixture: %v", err))
+	}
+	msg := json.RawMessage(b)
+	return &msg
+}
+
+func (s *SystemdTestSuite) TestLogPIDWithNonTrivialKeyValues(c *C) {
+	l1 := Log{
+		"_PID": mustJSONMarshal([]string{}),
+	}
+	l2 := Log{
+		"_PID": mustJSONMarshal(6),
+	}
+	l3 := Log{
+		"_PID": mustJSONMarshal([]string{"pid1", "pid2", "pid3"}),
+	}
+	l4 := Log{
+		"SYSLOG_PID": mustJSONMarshal([]string{"pid1", "pid2", "pid3"}),
+	}
+	l5 := Log{
+		"_PID":       mustJSONMarshal("42"),
+		"SYSLOG_PID": mustJSONMarshal([]string{"pid1", "pid2", "pid3"}),
+	}
+	l6 := Log{
+		"_PID":       mustJSONMarshal([]string{"42"}),
+		"SYSLOG_PID": mustJSONMarshal([]string{"pid1", "pid2", "pid3"}),
+	}
+	l7 := Log{
+		"_PID":       mustJSONMarshal([]string{"42", "42"}),
+		"SYSLOG_PID": mustJSONMarshal([]string{"singlepid"}),
+	}
+	c.Check(Log{}.PID(), Equals, "-")
+	c.Check(l1.PID(), Equals, "-")
+	c.Check(l2.PID(), Equals, "-")
+	c.Check(l3.PID(), Equals, "-")
+	c.Check(l4.PID(), Equals, "-")
+	// things starting with underscore are "trusted", so we trust
+	// them more than the user-settable ones:
+	c.Check(l5.PID(), Equals, "42")
+	c.Check(l6.PID(), Equals, "42")
+	c.Check(l7.PID(), Equals, "singlepid")
+}
+
+func (s *SystemdTestSuite) TestLogsMessageWithNonUniqueKeys(c *C) {
+	l1 := Log{
+		"MESSAGE": mustJSONMarshal([]string{"m1", "m2", "m3"}),
+	}
+	l2 := Log{
+		"MESSAGE": mustJSONMarshal("m1"),
+	}
+	l3 := Log{
+		// this is just "hello" in ascii
+		"MESSAGE": mustJSONMarshal([]rune{104, 101, 108, 108, 111}),
+	}
+	l4 := Log{
+		"MESSAGE": mustJSONMarshal(5),
+	}
+	c.Check(Log{}.Message(), Equals, "-")
```

Comment:
Can you add this entry too (`\r` at the end of the message from the bug report):
```
l5 := Log{ "MESSAGE": mustJSONMarshal([]rune{13})
c.Check(l5.Message(), Equals, "\r")
```

Reply:
sure

Reply:
done